lua/nvim-treesitter/ts_utils.lua

```diff
@@ -286,6 +286,20 @@ end
 -- @param col A column (0-based)
 function M.is_in_node_range(node, line, col)
   local start_line, start_col, end_line, end_col = node:range()
+  return M.is_in_range({
+    start_line,
+    start_col,
+    end_line,
+    end_col,
+  }, line, col)
+end
+
+--- Determines whether (line, col) position is in a range
+-- @param range Table with range {start_line, start_col, end_line, end_col}
+-- @param line A line (0-based)
+-- @param col A column (0-based)
+function M.is_in_range(range, line, col)
```

Comment:
Can this be marked private?

Reply:
Sorry for my ignorance but what does `@private` imply? :) Does it mean private to the module/plugin or something else? :)

Reply:
Plugins use the functions defined here, so we need to be mindful about what we expose here.

Thinking about it, can we make this a local function?

Reply:
Ah, actually the reason why I extracted it, was to be able to reuse it in a plugin. The reason was that I want to use the offset range and not the range of the node.

But maybe all of this should be done differently. As mentioned in https://github.com/neovim/neovim/pull/20158#issuecomment-1243321377 after sleeping on this, I'm wondering if, also here, the better approach (at least for the usecase I had) would be to support a fourth optional argument to eg is_in_node_range which is the metadata of the node. If this is not nil and has an entry range, this is then used instead of `node:range()`?

Then one could do for example:

```lua
for id, node, metadata in query:iter_captures(tree:root(), bufnr, first, last) do
  if is_in_node_range(node, line, col, metadata) then -- then takes #offset! into account
    ...
  end
end
```

How does that sound? Then the user does not need to explicitly touch the ranges anyway.

Reply:
I do prefer that. Ideally metadata should follow the node around and we should try to avoid anything internal about either of these.

Reply:
I guess `iter_captures` essentially already does this, it just doesn't encapsulate these fields, but maybe creating actual objects for these (and documenting them) is worth exploring.

Eventually we may end up moving a bunch of this to C, so maybe adding the object abstractions is a good pre-step.

@clason

Reply:
No, `capture` is a very different thing from a `node` and needs to remain distinct. Adding metadata to captures (like we already do with priorities) would make sense, though.

Reply:
We could turn `Capture` into an object, although I'm not sure what this would gain us. (I'm not too keen on adding more OO to our Lua library than we already have...)

Reply:
For the sake of completeness: `nodes` in vim.treesitter are references to lib-tree-sitter C objects. They do not know anything about metadata that _we_ add.

Reply:
> No, capture is a very different thing from a node and needs to remain distinct.

Agreed!

> Adding metadata to captures (like we already do with priorities) would make sense, though.

Just to make sure I understand, if the "capture" is not a concept on its on (eg a table with `node` and `metadata` keys), what do you mean when you say adding metadata to captures?

Reply:
Capture is _already_ a concept on its own, different from `node`.

Reply:
What we're talking about is to formalize a `Capture` object that is a table that contains
* the `capture_id`
* the `capture_name` (e.g., `@keyword`)
* the `captured_node` (e.g., `"return"`)
* the `metadata` (which is itself a table `{ priority = ..., offset = ..., ... }`)

Reply:
All this is a discussion for core, though.

Reply:
`range` is already returned by the iterator as part of the metadata

Reply:
You can see this by for example running
```lua
local ts = vim.treesitter
local bufnr = vim.fn.bufnr()
local tree = ts.get_parser(bufnr, 'lua'):parse()[1]

local cquery = ts.parse_query('lua', [[((string) @test (#offset! @test 0 1 0 -1))]])
for pattern, match, metadata in cquery:iter_matches(tree:root(), bufnr, 0, -1) do
  for id, node in pairs(match) do
    print(vim.inspect(metadata[id]))  -- prints eg {range = { 4, 38, 4, 82 }}
  end
end
```

Reply:
Yes, I believe you -- I myself mentioned that `range` is added as `metadata` (only) by the `offset!` directive.

Reply:
Right, I see!

My suggestion for now would then be to update this PR to:
* Make `is_in_range` local (used by both `is_in_node_range` and `is_in_capture_range` below).
* Add `is_in_capture_range(capture, line, col)`.
* Rename `get_match_text`/`get_match_range` that's added here to `get_capture_text`/`get_capture_range`.

The argument `capture` to these functions would be a table of the form `{node = ..., metadata = ...}` which is already the form of the values of the tables `get_matches` return.

Then a more longterm discussion could be had in core, where similar functions may be added in the future with a similar signature (or not).

If an optional `opts` is added to `get_node_text` that may contain metadata in https://github.com/neovim/neovim/pull/20158, then `get_capture_text` here can later be updated to make use of this.

Reply:
> Then a more longterm discussion could be had in core,

_needs_ to be had in core (after 0.8). we can't have fundamental API diverging. For this reason I am also hesitant to make such big changes here.

Basically,

* it is OK to refactor functions that are only here.
* functions that have been upstreamed (and will be kept) will be removed here, so any _changes_ should now go to core
* new functions should in general be added here first.
* we need to be very careful to use consistent naming, and be clear on which level we operate (node, capture, match -- if that is even needed). in particular, nodes do not and should not have their range corrected!

Reply:
Makes sense!

> needs to be had in core (after 0.8). 

:+1: 

> it is OK to refactor functions that are only here.

it would only internally change `is_in_node_range` but not its behaviour

> functions that have been upstreamed (and will be kept) will be removed here, so any changes should now go to core

we won't be doing this

> new functions should in general be added here first.

there would be three new functions here `is_in_capture_range`, `get_capture_range` and `get_capture_text`

> we need to be very careful to use consistent naming, and be clear on which level we operate (node, capture, match -- if that is even needed). in particular, nodes (which belong to the tree-sitter library) do not and should not have their range corrected! I also need to check the code carefully whether there is a meaningful distinction between match and capture (which then needs to be kept).

my understanding is that a "match" is a match of a query which can contain multiple captures. this aligns well with eg `get_matches` here and `iter_matches` in core. `get_matches` returns a list of matches, each of which can have more than one "capture" (keyed by the capture-name). For example if you call `get_matches(0, 'injections')` on a markdown file with code-blocks you might get a something of the form:

```lua
{
  ...
  {  -- match with two captures called content and language
    content = {node = <ts_node>, metadata = {...}},
    language = {node = <ts_node>, metadata = {...}},
  }
  ...
  {  -- match with one capture called html
    html = {node = <ts_node>, metadata = {...}},
  }
  ...
}
```

Reply:
I've updated the PR with the above now